File: CLDemo-OC/CLDemo-OC/Class/Base/CLTabbarController/View/CLCustomTabbar.m
//
//  CustomTabbar.m
//  CLDemo
//
//  Created by JmoVxia on 2016/11/17.
//  Copyright © 2016年 JmoVxia. All rights reserved.
//

#import "CLCustomTabbar.h"
#import "UIImage+CLScaleToSize.h"
#import "UIButton+CLBlockAction.h"
#import "UIView+CLSetRect.h"

@interface CLCustomTabbar ()

@property (nonatomic,weak) UIButton *bulgeButton;

@end

@implementation CLCustomTabbar

- (UIButton *) bulgeButton {
    if (_bulgeButton == nil) {
        UIImage *normalImage = [UIImage originImage:[UIImage imageNamed:@"tabBar_publish_icon"] scaleToSize:CGSizeMake(80, 80)];
        UIImage *selectedImage = [UIImage originImage:[UIImage imageNamed:@"tabBar_publish_click_icon"] scaleToSize:CGSizeMake(80, 80)];
        UIButton * button = [[UIButton alloc] init];
        [button setImage:selectedImage forState:UIControlStateSelected];
        [button setImage:normalImage forState:UIControlStateNormal];
        __weak __typeof(self) weakSelf = self;
        [button addActionBlock:^(UIButton *button) {
            __typeof(&*weakSelf) strongSelf = weakSelf;
            [strongSelf bulgeButtonAction:button];
        }];
        [self addSubview:button];
        _bulgeButton = button;
    }
    return _bulgeButton;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setTranslucent:NO];
    }
    return self;
}

-(void)layoutSubviews {
    [super layoutSubviews];
    int index = 0;
    CGFloat itemW = self.cl_width/5.0;

    for (UIView *subviews in self.subviews) {
        //取到系统tabbar的Item方法
        if ([@"UITabBarButton" isEqualToString:NSStringFromClass(subviews.class)]) {
            if (index == 2) {
                subviews.hidden = YES;
            }
            index++;
        }
    }
    self.bulgeButton.frame = CGRectMake(0, 0, itemW, itemW);
    self.bulgeButton.center = CGPointMake(self.cl_width/2.0, (self.cl_height - 30 - cl_safeBottomMargin)/2.0);
}
//判断点是否在响应范围
- (BOOL)pointInside:(CGPoint)point withEvent:(UIEvent *)event {
    if (self.isHidden == NO) {
        UIBezierPath *circle  = [UIBezierPath bezierPathWithArcCenter:self.bulgeButton.center radius:35 startAngle:0 endAngle:2* M_PI clockwise:YES];
        UIBezierPath *tabbar  = [UIBezierPath bezierPathWithRect:self.bounds];
        if ( [circle containsPoint:point] || [tabbar containsPoint:point]) {
            return YES;
        }
        return NO;
    }else {
        return [super pointInside:point withEvent:event];
    }
}

- (void)bulgeButtonAction:(UIButton *)button {
    if (self.bulgeCallBack) {
        self.bulgeCallBack();
    }
    NSLog(@"-------->>凸起按钮被点击了");
}

@end

File: CLDemo-Swift/CLDemo-Swift/Class/Base/CLTabbarController/View/CLCustomTabbar.swift
//
//  CLCustomTabbar.swift
//  CKD
//
//  Created by JmoVxia on 2020/6/12.
//  Copyright © 2020 JmoVxia. All rights reserved.
//

import UIKit

class CLCustomTabbar: UITabBar {
    var bulgeCallBack: ((UITabBarItem) -> ())?
    private lazy var bulgeButton: UIButton = {
        let view = UIButton()
        view.setImage(UIImage(named: "recording"), for: .normal)
        view.setImage(UIImage(named: "recording"), for: .selected)
        view.setImage(UIImage(named: "recording"), for: .highlighted)
        view.addTarget(self, action: #selector(bulgeButtonAction), for: .touchUpInside)
        return view
    }()
    private lazy var lineView: UIView = {
        let view = UIView()
        view.backgroundColor = .hex("EEEEEE")
        return view
    }()
    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .white
        backgroundImage = UIImage()
        shadowImage = UIImage()
        tintColor = .hex("#666666")
        isTranslucent = false
        addSubview(lineView)
        addSubview(bulgeButton)
    }
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}
extension CLCustomTabbar {
    override func layoutSubviews() {
        super.layoutSubviews()
        let width = bounds.width
        let height = bounds.height
        var index = 0
        for subviews in subviews {
            if ("UITabBarButton" == NSStringFromClass(type(of: subviews).self)) {
                if index == 2 {
                    subviews.isHidden = true
                }
                index += 1
            }
        }
        bulgeButton.frame = CGRect(x: 0, y: 0, width: 81, height: 60)
        bulgeButton.center = CGPoint(x: width * 0.5, y: (height - safeAreaEdgeInsets.bottom) * 0.5 - 5)
        lineView.frame = CGRect(x: 0, y: 0, width: width, height: 0.5)
    }
}
extension CLCustomTabbar {
    ///判断点是否在响应范围
    override func point(inside point: CGPoint, with event: UIEvent?) -> Bool {
        let circle = UIBezierPath(arcCenter: bulgeButton.center, radius: 28, startAngle: 0, endAngle: .pi * 2, clockwise: true)
        let tabbar = UIBezierPath(rect: bounds)
        if circle.contains(point) || tabbar.contains(point) {
            return true
        }
        return super.point(inside: point, with: event)
    }
}
@objc extension CLCustomTabbar {
    func bulgeButtonAction() {
        guard let item = items?[safe: 2] else { return }
        bulgeCallBack?(item)
        let animation = CASpringAnimation(keyPath: "transform.scale")
        animation.mass = 0.6
        animation.stiffness = 80
        animation.damping = 10
        animation.initialVelocity = 0.5
        animation.duration = 0.5
        animation.fromValue = 0.25
        animation.toValue = 1
        animation.timingFunction = CAMediaTimingFunction(controlPoints: 0.6, 0.1, 0.30, 0.90)
        bulgeButton.layer.add(animation, forKey: nil)
    }
}
extension CLCustomTabbar {
    ///显示小红点
    func showBadgeOnItem(index: Int, badgeValue: Int) {
        guard let items = items, items.count > index else {
            return
        }
        let item = items[index]
        if badgeValue > 99 {
            item.badgeValue = "99+"
        }else {
            item.badgeValue = "\(badgeValue)"
        }
    }
    ///移除小红点
    func hiddenBadgeOnItem(index: Int) {
        guard let items = items, items.count > index else {
            return
        }
        let item = items[index]
        item.badgeValue = nil
    }
}
